packages/server/src/mapping/convmap/service.ts

```diff
@@ -97,6 +104,19 @@ export class ConvmapService extends Service {
     return convmaps
   }
 
+  async map(tunnelId: uuid, threadId: uuid, userId: uuid): Promise<Convmap> {
+    const convmap = await this.getByThreadId(tunnelId, threadId)
+    if (convmap) {
+      return convmap
+    }
+
+    return this.barrier.once(tunnelId, threadId, async () => {
+      const tunnel = await this.tunnels.get(tunnelId)
+      const conversation = await this.conversations.create(tunnel!.clientId, userId)
+      return this.create(tunnelId, conversation.id, threadId)
+    })
+  }
+
```

Comment:
@laurentlp here's a barrier in use. The code in the callback will be excuted only once, and return the same value for all Promises waiting on it. The way a single "task" is identified is with arguments before the callback. This particular barrier is a 2d barrier (similarly to our 2d caches), so the key is composed of two keys.